### PR TITLE
Migrating from management SDK to autogenerated C# code for deployments and related cmdlets

### PR DESCRIPTION
**Pull request by Xynoclafe**
## Description

* Replace Resources management SDK with auto generated C# source code based on swagger API for deployments and related cmdlets.
* Update API version to 2022-09-01 from 2021-04-01
* Split ResourceManagerSDKClient and a few other classes into multiple files to support both the old SDK and autogenerated code side-by-side.

## Checklist

- [x] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** select appropriate branch. Cmdlets from Autorest.PowerShell should go to [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch. 
* **SHOULD** make the title of PR clear and informative, and in the present imperative tense. 
* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. Add changelog in description section if PR goes into [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** have approved design review for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) ([_Microsoft internal only_](../blob/main/CONTRIBUTING.md#onboarding)) with following situations
    * Create new module from scratch
    * Create new resource types which are not easy to conform to [Azure PowerShell Design Guidelines](../blob/main/documentation/development-docs/design-guidelines)
    * Create new resource type which name doesn't use module name as prefix
    * Have design question before implementation
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** introduce [breaking changes](../blob/main/documentation/breaking-changes/breaking-changes-definition.md) in Az minor release except preview version.
* **SHOULD NOT** adjust version of module manually in pull request

### CODE REVIEW

File: src/Resources/Resources.Sdk/generate.ps1 [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+# # Generate package with policy tag
+# Start-AutoRestCodeGeneration -ResourceProvider "resources/resource-manager" -SdkRepoRootPath "$PSScriptRoot\..\..\..\.." -AutoRestVersion "v2" -AutoRestCodeGenerationFlags "--tag=package-policy-2021-06" -SdkGenerationDirectory "$PSScriptRoot\GeneratedPolicy202106"
+
+# Generate package with deployment scripts
+# Start-AutoRestCodeGeneration -ResourceProvider "resources/resource-manager" -SdkRepoRootPath "$PSScriptRoot\..\..\..\.." -AutoRestVersion "v2" -AutoRestCodeGenerationFlags "--tag=package-deploymentscripts-2019-10-preview" -SdkGenerationDirectory "$PSScriptRoot\Generated\DeploymentScripts"
+
+# Generate package with privatelink tag
+Start-AutoRestCodeGeneration -ResourceProvider "resources/resource-manager" -SdkRepoRootPath "$PSScriptRoot\..\..\..\.." -AutoRestVersion "v2" -AutoRestCodeGenerationFlags "--tag=package-privatelinks-2020-05" -SdkGenerationDirectory "$PSScriptRoot\Generated\PrivateLinks"
```
[the remark this answers is not in the document]
Comment (Xynoclafe) [reply src/Resources/Resources.Sdk/generate.ps1:21]: These were the default states of the files I copied over from the sdk repo. Not sure if the file is even required in this repo, but kept it around at least for the first iteration

File: src/Resources/ResourceManager/ResourceManager.csproj [hunk 2]
```diff
@@ -12,9 +12,12 @@
   </PropertyGroup>
 
   <ItemGroup>
-    
-    <PackageReference Include="Microsoft.Azure.Management.ResourceManager" Version="3.17.3-preview" />
```
Comment (msJinLei) [inline src/Resources/ResourceManager/ResourceManager.csproj:16]: This is unnecessary change

File: src/Resources/Tags/Tags.csproj [hunk 3]
```diff
@@ -12,7 +12,7 @@
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.Azure.Management.ResourceManager" Version="3.17.3-preview" />
+	<PackageReference Include="Microsoft.Azure.Management.ResourceManager" Version="3.17.3-preview" />
```
Comment (msJinLei) [inline src/Resources/Tags/Tags.csproj:15]: Unnecessary change

**Review by msJinLei — CHANGES_REQUESTED** [review]

File: src/Resources/ResourceManager/ResourceManager.csproj [hunk 4]
```diff
@@ -12,11 +12,14 @@
...
 
   <ItemGroup>
-    
     <PackageReference Include="Microsoft.Azure.Management.ResourceManager" Version="3.17.3-preview" />
     <PackageReference Include="Microsoft.Extensions.Caching.Memory" Version="2.2.0" />
   </ItemGroup>
 
+  <ItemGroup>
+	<ProjectReference Include="..\Resources.Sdk\Resources.Sdk.csproj" />
+  </ItemGroup>
```
Comment (msJinLei) [inline src/Resources/ResourceManager/ResourceManager.csproj:21]: As you add generated sdk project, why still need Microsoft.Azure.Management.ResourceManager?

**Review by msJinLei — COMMENTED** [review]
As your PR contains no breaking change, can it be released in next sprint?

File: src/Resources/ResourceManager/ResourceManager.csproj (hunk 4 above)
Comment (Xynoclafe) [reply src/Resources/ResourceManager/ResourceManager.csproj:21]: Az.Resources has cmdlets for multiple resource types owned and supported by different teams. Since we don't have ownership of the entire module nor do we have a complete understanding of the other cmdlets, updating them runs a high risk of breaking functionality for Azure customers. 
We are using the checked in SDK code with only the cmdlets that we own while the rest of the cmdlets still use the older SDK. 

Moreover, the TestFx project also uses the old SDK, making it harder to completely remove all references.

File: src/Resources/ResourceManager/ResourceManager.csproj [hunk 5]
```diff
@@ -12,11 +12,14 @@
   </PropertyGroup>
 
   <ItemGroup>
-    
     <PackageReference Include="Microsoft.Azure.Management.ResourceManager" Version="3.17.3-preview" />
```
[the remark this answers is not in the document]
Comment (Xynoclafe) [reply src/Resources/ResourceManager/ResourceManager.csproj:15]: Az.Resources has cmdlets for multiple resource types owned and supported by different teams. Since we don't have ownership of the entire module nor do we have a complete understanding of the other cmdlets, updating them runs a high risk of breaking functionality for Azure customers. 
We are using the checked in SDK code with only the cmdlets that we own while the rest of the cmdlets still use the older SDK. 

Moreover, the TestFx project also uses the old SDK, making it harder to completely remove all references.

**Comment by Xynoclafe** [comment]
> As your PR contains no breaking change, can it be released in next sprint?

That is fine. Out of curiosity, what are the milestone dates for the next release?

**Review by msJinLei — APPROVED** [review]

**Review by isra-fel — DISMISSED** [review]
For `Resources.Sdk`, if you [follow the guide](https://eng.ms/docs/cloud-ai-platform/azure-core/azure-management-and-platforms/control-plane-bburns/azure-cli-tools-azure-cli-powershell-and-terraform/azure-cli-tools/devguide/azps/coding/generate-service-client-code#generate-service-client-code-for-owned-rp), there should be a `README.md` checked in.
It should contian the information about swagger, basically the urls to the swagger files. That file is useful for diagnostics and future maintenance so please do check it in.

**Comment by vidai-msft** [comment]
@Xynoclafe please resolve the conflicts.

**Comment by Xynoclafe** [comment]
> @Xynoclafe please resolve the conflicts.

They should be resolved now

**Comment by Xynoclafe** [comment]
> Hello @isra-fel - any updates on review here?

We have some failures to resolve because of the namespace change

**Comment by vidai-msft** [comment]
@Xynoclafe you can follow the guidance here https://github.com/Azure/azure-powershell/blob/main/documentation/Debugging-StaticAnalysis-Errors.md#breaking-changes to suppress the static analysis issues?

**Review by vidai-msft — APPROVED** [review]

**Comment by isra-fel** [comment]
The breaking change issues were due to the migration of SDK. Although the namespaces and names of the previous type and the current type are the same, they are from different  assemblies, that's why they were identified as different types.
Anyway I don't think this would affect script users, so we can consider them not breaking changes.
I approve the PR. @vidai please go ahead.
Thanks.
